thumbnail-service/src/test/java/no/sikt/nva/thumbnail/ThumbnailRequestHandlerTest.java

```diff
@@ -1,28 +1,205 @@
+        handler = new ThumbnailRequestHandler(s3Client);
+        assertThrows(RuntimeException.class, () -> handler.handleRequest(s3Event, CONTEXT));
+        assertThat(appender.getMessages(), containsString(expectedMessage));
+    }
 
+    @Test
+    void shouldThrowExceptionWhenMimeTypeIsNotSupported() throws IOException {
+        var s3Event = createNewFileUploadEvent(UNSUPPORTED_FILES_PATH + "/" + ZIP_FILE, ZIP_MIME_TYPE, ZIP_FILE);
+        assertThrows(RuntimeException.class, () -> handler.handleRequest(s3Event, CONTEXT));
+        assertThat(appender.getMessages(), containsString(COULD_NOT_CREATE_THUMBNAIL_LOG_MESSAGE));
     }
 
     @Test
-    void shouldReturnHandleUriWhenInputIsValidUri() {
-        handler.handleRequest(null, context);
+    void shouldThrowExceptionWhenS3ClientWontPutObject() throws IOException {
+        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + JPEG_FILE, JPEG_MIME_TYPE, JPEG_FILE);
+        var expectedMessage = randomString();
+        s3PutObjectThrowsException(expectedMessage);
+        assertThrows(RuntimeException.class, () -> handler.handleRequest(s3Event, CONTEXT));
+        assertThat(appender.getMessages(), containsString(expectedMessage));
+    }
+
+    @Test
+    void shouldUploadNativeImageToS3WhenNoExceptionOccurs() throws IOException {
+        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + JPEG_FILE, JPEG_MIME_TYPE, JPEG_FILE);
+        when(s3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(
+            PutObjectResponse.builder().build());
+        verify(s3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
+        assertDoesNotThrow(() -> handler.handleRequest(s3Event, CONTEXT));
+    }
+
+    @Test
+    void shouldHandleIdenticalSizedThumbnails() throws IOException {
+        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + ALREADY_HAVE_SIZE_AS_THUMBNAIL_OUTPUT, PNG_MIME_TYPE,
+                                               ALREADY_HAVE_SIZE_AS_THUMBNAIL_OUTPUT);
+        when(s3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(
+            PutObjectResponse.builder().build());
+        verify(s3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
+        assertDoesNotThrow(() -> handler.handleRequest(s3Event, CONTEXT));
+    }
+
+
+    @Test
+    void shouldHandleTinySizedThumbnails() throws IOException {
+        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + TINY_IMAGE, PNG_MIME_TYPE,
+                                               TINY_IMAGE);
+        when(s3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(
+            PutObjectResponse.builder().build());
+        verify(s3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
+        assertDoesNotThrow(() -> handler.handleRequest(s3Event, CONTEXT));
+    }
+
```

Comment:
These tests succeed when we comment out the last line. This means that they only test that the handler does not throw an exception which means that an empty handler would suffice for these tests to pass.  The only reason the build is failing is because of the Jacoco coverage. 

What I mean is that these tests pass:
```
    @Test
    void shouldUploadNativeImageToS3WhenNoExceptionOccurs() throws IOException {
        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + JPEG_FILE, JPEG_MIME_TYPE, JPEG_FILE);
        when(s3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(
            PutObjectResponse.builder().build());
        verify(s3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
//        assertDoesNotThrow(() -> handler.handleRequest(s3Event, CONTEXT));
    }

    @Test
    void shouldHandleIdenticalSizedThumbnails() throws IOException {
        var s3Event = createNewFileUploadEvent(IMAGES_PATH + "/" + ALREADY_HAVE_SIZE_AS_THUMBNAIL_OUTPUT, PNG_MIME_TYPE,
                                               ALREADY_HAVE_SIZE_AS_THUMBNAIL_OUTPUT);
        when(s3Client.putObject(any(PutObjectRequest.class), any(RequestBody.class))).thenReturn(
            PutObjectResponse.builder().build());
        verify(s3Client, times(1)).putObject(any(PutObjectRequest.class), any(RequestBody.class));
//        assertDoesNotThrow(() -> handler.handleRequest(s3Event, CONTEXT));
    }
```

Reply:
There fixed, test should have failed since the verify was before the handleRequest call, but since placing the input file inside the s3Client creates an putObject request, I didn't catch this. When I had made the test and before implementing the method the handler was throwing exceptions.

Reply:
But as you can see from my top comment, I would like suggestion how to create a good unit test for this.